vortex-duckdb/src/exporter/mod.rs

```diff
@@ -208,18 +210,186 @@ impl VectorExt for Vector {
                 true
             }
             Mask::Values(arr) => {
-                // TODO(joe): do this MUCH better, with a shifted u64 copy
-                let mut null_count = 0;
-                // SAFETY: Caller guaranteees this.
-                let validity = unsafe { self.ensure_validity_slice(len) };
-                for (idx, v) in arr.boolean_buffer().slice(offset, len).iter().enumerate() {
-                    if !v {
-                        validity.set(idx, false);
-                        null_count += 1;
+                let arr_bits: &[u64] = {
+                    let byte_slice = arr.boolean_buffer().inner().as_slice();
+                    unsafe {
+                        std::slice::from_raw_parts(
+                            byte_slice.as_ptr() as _,
+                            byte_slice.len().div_ceil(size_of::<u64>()),
+                        )
+                    }
+                };
+                let sliced_bits = &arr_bits.view_bits::<Lsb0>()[offset..][..len];
+                let true_count = sliced_bits.count_ones();
+                if true_count == len {
+                    if let Some(validity) = unsafe { self.validity_slice_mut(len) } {
+                        validity.fill(true);
                     }
+                } else if true_count == 0 {
+                    unsafe { self.ensure_validity_slice(len) }.fill(false);
+                } else {
+                    // SAFETY: Caller guaranteees this.
+                    let validity = unsafe { self.ensure_validity_slice(len) };
+                    validity[..len].copy_from_bitslice(sliced_bits);
```

Comment:
do you need [..len]?

Reply:
Looks like I do not. I was being defensive

Reply:
You do need the len here since values of the mask are sliced `mask[offset..][..len]` and we request len so validity here will be offset bits longer